pay-api/src/pay_api/services/eft_refund.py

```diff
@@ -0,0 +1,198 @@
+                        link_group_id=link_group_id).flush()
+                    # if corp_type := CorpTypeModel.find_by_code(invoice.corp_type_code):
+                    #     if corp_type.has_partner_disbursements:
+                    #         reversal_total += cil.amount
+
+                # if reversal_total > 0:
+                #     PartnerDisbursementsModel(
+                #         amount=reversal_total,
+                #         is_reversal=True,
+                #         partner_code=invoice.corp_type_code,
+                #         status_code=DisbursementStatus.WAITING_FOR_JOB.value,
+                #         target_id=invoice.id,
+                #         target_type=EJVLinkType.INVOICE.value
+                #     ).flush()
+
+        current_balance = EFTCreditModel.get_eft_credit_balance(latest_eft_credit.short_name_id)
+        if existing_balance != current_balance:
+            short_name_history = EFTHistoryModel.find_by_related_group_link_id(latest_link.link_group_id)
+            EFTHistoryService.create_invoice_refund(
+                EFTHistoryModel(short_name_id=latest_eft_credit.short_name_id,
+                                amount=invoice.total,
+                                credit_balance=current_balance,
+                                payment_account_id=payment_account.id,
+                                related_group_link_id=link_group_id,
+                                statement_number=short_name_history.statement_number if short_name_history else None,
+                                invoice_id=invoice.id,
+                                is_processing=True,
+                                hidden=False)).flush()
+
+        return InvoiceStatus.REFUND_REQUESTED.value
+
+    @staticmethod
+    def refund_eft_credits(shortname_id: int, amount: str):
+        """Refund the amount to eft_credits table based on short_name_id."""
+        refund_amount = Decimal(amount)
+        eft_credits = EFTCreditModel.get_eft_credits(shortname_id)
+        eft_credit_balance = EFTCreditModel.get_eft_credit_balance(shortname_id)
+
+        if refund_amount > eft_credit_balance:
+            raise BusinessException(Error.INVALID_REFUND)
+
+        for credit in eft_credits:
+            if refund_amount <= 0:
+                break
+            credit_amount = Decimal(credit.remaining_amount)
+            if credit_amount <= 0:
+                continue
+
+            deduction = min(refund_amount, credit_amount)
+            credit.remaining_amount -= deduction
+            refund_amount -= deduction
+
+            credit.save()
+
+    @staticmethod
+    def update_shortname_refund(refund_id: int, data: EFTShortNameRefundPatchRequest) -> EFTRefundModel:
+        """Update the refund status."""
+        refund = EFTRefundModel.find_pending_refund(refund_id)
+        refund.comment = data.comment
+        refund.decline_reason = data.declined_reason
+        refund.status = data.status
```

Comment:
We will need some additional logic when declining. 
1) The funds removed from the credits will have to be put back.
2) History record status needs to be updated - amount can stay the same for history purposes and the UI will automatically display what is relevant for history status.

Feel free to just add # TODO and I can follow up on this

Reply:
whoops forgot, I'll add it in

Reply:
Should be updated now

Reply:
EFTCredit.get_eft_credits only returns credits with a remaining amount this will probably be an issue, might need to add a flag for that query so it returns zeroed out remaining amounts.

We should also make sure we don't add more then the amount originally on the credit since it ties back to a TDI17 transaction.

Reply:
sounds good, remaining vs amount check

will add in flag to include 0 amount credits when a refund is "DECLINED" (previously rejected)

Reply:
Going to write a separate function, it gets fairly confusing with reversals and negative amounts.

Reply:
Updated with unit tests